openscm/timeframes.py

```diff
@@ -0,0 +1,513 @@
+
+    Parameters
+    ----------
+    source_len
+        Length of source data values
+    source
+        Source timeframe
+    target
+        Target timeframe
+
+    Returns
+    -------
+    _Interpolation
+        Object wrapping the actual interpolation points and additional information
+    """
+    linearization_points_len = 2 * source_len + 1
+    # Use double the time points to stick to integers when using half the source period
+    # length:
+    source = Timeframe(2 * source.start_time, 1 * source.period_length)
+    target = Timeframe(2 * target.start_time, 2 * target.period_length)
+
+    if target.start_time >= source.start_time:
+        skip_len = 1 + (target.start_time - source.start_time) // source.period_length
+        linearization_points_len -= skip_len
+        source.start_time += skip_len * source.period_length
+        first_target_point = target.start_time
+        first_interval_length = target.period_length
+    else:
+        first_target_point = target.start_time + target.period_length
+        first_interval_length = source.start_time - target.start_time
+
+    source_stop_time = source.get_stop_time(linearization_points_len)
+    target_len = 1 + (source_stop_time - first_target_point) // target.period_length
+    target_stop_time = target.get_stop_time(target_len)
+
+    if source_stop_time > target_stop_time:
+        last_interval_length = source_stop_time - target_stop_time
+    else:
+        last_interval_length = target.period_length
+
+    interpolation_points, indices = np.unique(
+        np.concatenate(
+            (target.get_points(target_len), source.get_points(linearization_points_len))
+        ),
+        return_index=True,
+    )
+    target_indices = np.where(indices < target_len)[0]
+
+    if source_stop_time <= target_stop_time:
+        target_indices = target_indices[:-1]
+
+    return _Interpolation(
+        data_len=source_len,
+        first_interval_length=first_interval_length // 2,
+        last_interval_length=last_interval_length // 2,
+        period_length=target.period_length // 2,
+        points=interpolation_points / 2,
+        target_indices=target_indices,
+    )
+
+
+def _calc_interval_averages(
+    interpolation: _Interpolation, interpolation_values: np.ndarray
+) -> np.ndarray:
+    """
+    Calculate the average value for each period (i.e. integral over the period divided
+    by the period length) for the "interpolated" values (see
+    :func:`openscm.timeframes._calc_interpolation_points`) array
+    ``interpolation_values`` at the time points ``interpolation_points``.
+
+    Parameters
+    ----------
+    interpolation
+        Object wrapping the interpolation points in time ("x-values" corresponding to
+        ``interpolation_values``) and additional information
+    interpolation_values
+        Interpolation values ("y-values")
+
+    Returns
+    -------
+    np.ndarray
+        Array of the interval/period averages
+    """
```

Comment:
Is there not a nicer way to write `int = (left_edge^2 - right_edge^2) / 2?

If not, can we add a comment like 'Simply calculating integral, taking care of shaping arrays correctly is fiddly'? Or split the calculation step and reshaping steps? np.add.reduceat is confusing, putting all the calculation steps in there too makes debugging tricky...

Reply:
Should be clearer now together with your suggestion about the trapezium rule